homeassistant/components/owntracks/__init__.py

```diff
@@ -0,0 +1,219 @@
+            _LOGGER.error("Unable to parse payload as JSON: %s", payload)
+            return
+
+        message['topic'] = topic
+        hass.helpers.dispatcher.async_dispatcher_send(
+            DOMAIN, hass, context, message)
+
+    await hass.components.mqtt.async_subscribe(
+        context.mqtt_topic, async_handle_mqtt_message, 1)
+
+    return True
+
+
+async def handle_webhook(hass, webhook_id, request):
+    """Handle webhook callback."""
+    context = hass.data[DOMAIN]['context']
+    message = await request.json()
+
+    # Android doesn't populate topic
+    if 'topic' not in message:
+        headers = request.headers
+        user = headers.get('X-Limit-U')
+        device = headers.get('X-Limit-D', user)
+
+        if user is None:
+            _LOGGER.warning('Set a username in Connection -> Identification')
+            return json_response(
+                {'error': 'You need to supply username.'},
+                status=400
+            )
+
+        topic_base = re.sub('/#$', '', context.mqtt_topic)
+        message['topic'] = '{}/{}/{}'.format(topic_base, user, device)
+
+    hass.helpers.dispatcher.async_dispatcher_send(
+        DOMAIN, hass, context, message)
+    return json_response([])
+
+
+class OwnTracksContext:
+    """Hold the current OwnTracks context."""
+
+    def __init__(self, hass, secret, max_gps_accuracy, import_waypoints,
+                 waypoint_whitelist, region_mapping, events_only, mqtt_topic):
+        """Initialize an OwnTracks context."""
+        self.hass = hass
+        self.secret = secret
+        self.max_gps_accuracy = max_gps_accuracy
+        self.mobile_beacons_active = defaultdict(set)
+        self.regions_entered = defaultdict(list)
+        self.import_waypoints = import_waypoints
+        self.waypoint_whitelist = waypoint_whitelist
+        self.region_mapping = region_mapping
+        self.events_only = events_only
+        self.mqtt_topic = mqtt_topic
+
+    @callback
+    def async_valid_accuracy(self, message):
+        """Check if we should ignore this message."""
+        acc = message.get('acc')
+
+        if acc is None:
+            return False
+
+        try:
+            acc = float(acc)
+        except ValueError:
+            return False
+
+        if acc == 0:
+            _LOGGER.warning(
+                "Ignoring %s update because GPS accuracy is zero: %s",
+                message['_type'], message)
+            return False
+
+        if self.max_gps_accuracy is not None and \
+                acc > self.max_gps_accuracy:
+            _LOGGER.info("Ignoring %s update because expected GPS "
+                         "accuracy %s is not met: %s",
+                         message['_type'], self.max_gps_accuracy,
+                         message)
+            return False
+
+        return True
+
+    async def async_see(self, **data):
+        """Send a see message to the device tracker."""
+        await self.hass.components.device_tracker.async_see(**data)
```

Comment:
This function doesn't seem to exist.

Reply:
If we will overwrite this method, maybe we should `raise NotImplementedError` here?

Reply:
https://github.com/home-assistant/home-assistant/pull/18777